### ISSUE TITLE
How to use with geopandas?

### ISSUE THREAD

**Issue by cmcconomyfwig** [hgrecco/pint-pandas#82]
I am interested in using this with Geopandas; are there are specific instructions to integrate pint / pint-pandas with geopandas?

**Comment by hgrecco** [comment]
My (very small) understanding of Geopandas is that is built in on top of pandas Series and DataFrame. So it should not be to hard to approach supporting geopandas in the same way we have done it for pandas. I am not a Geopandas user but I will be happy to support any attempt to make it work with pint.

**Comment by danieldjewell** [comment]
>built on top of pandas

Yes, I can confirm this... There's basically pd.Series >> GeoSeries and pd.DataFrame >> GeoDataFrame. And in fact GeoDataFrame inherits pd.DataFrame: https://github.com/geopandas/geopandas/blob/master/geopandas/geodataframe.py

The biggest difference with geopandas is the addition of "geometry" for a dataframe -- almost like a column that always exists. Usually stored as shapely objects... And with a bunch of helper functions to do things like set the CRS (coordinate reference system... WGS84 is one - used by GPS) and convert.

A challenge might be trying to get pint to interact with the geometry data... For example, it's possible to convert Latitude/Longitude into UTM

**Comment by hgrecco** [comment]
I think the interaction with geometry data should be step two, ... at least 1.5 

My proposal is:
1. I will create a new pint-geopandas repo similar to pint-pandas
2. I will grant commit privileges to the people that want to participate. Geopandas users are encouraged to do it!
3. My suggestion is that we write some usecases first and then convert them into tests, even for the geometry part. People without knowledge about geopandas but that undersdant Pint (like myself and the pint-pandas committers) will find it easiert to help once the required functionality is written as code
4. We code
5. We enjoy pint-geopandas

**Comment by danieldjewell** [comment]
> I think the interaction with geometry data should be step two, ... at least 1.5 

That's probably a good idea.... 

Based on my experience w/Geopandas, it *might* be possible to further extend the Geometry type (It's implemented similarly to pint-pandas with a pd.api.extensions.ExtensionDType and ExtensionArray)... In order to preserve the functionality, pint-geopandas would probably have to extend the already extended dtype/arraytype? (And there are some pretty big things implemented in the Geopandas extension types and it utilizes shapely/pygeos...) 

> My proposal is:
> 1. I will create a new pint-geopandas repo similar to pint-pandas
> 2. I will grant commit privileges to the people that want to participate. Geopandas users are encouraged to do it!

Great idea. Some of the geo stuff is ... different. I'm certainly no expert but I know enough to know that there are some complicated things w/geometry.

> 3. My suggestion is that we write some usecases first and then convert them into tests, even for the geometry part. People without knowledge about geopandas but that undersdant Pint (like myself and the pint-pandas committers) will find it easiert to help once the required functionality is written as code

Yeah, I think that's a good idea -- Maybe even make this number 1? Kinda sketch out the use cases and do some preliminary looking to see what the workload is to implement?

Because here's the thing:

* Geopandas geodataframes can easily be converted to regular dataframes (and back?) -- so while not ideal, pint-pandas could still be used
* There's a whole additional layer of serialization/deserialization in the Geo world - formats like GeoJSON and so on. Given that handling of formats like parquet for geopandas has been an issue (maybe it's better now) and still is an issue for pint-pandas.... That's going to be a potential challenge.
* What's the value of being able to have regular (non-geometry) columns be pint columns w/o tight integration of geometry? (I honestly don't know one way or the other.) 



> 4. We code
> 5. We enjoy pint-geopandas

**Comment by hgrecco** [comment]
@danieldjewell  @cmcconomyfwig 

I have created https://github.com/hgrecco/pint-geopandas

**Comment by redhog** [comment]
I might be late to the game here, but have a look at https://github.com/redhog/pint-proj/blob/master/docs/Example%20usage.ipynb which is something I threw together. It supports converting between geopandas point geometries and pint units that represent coordinates, as well as doing distance calculations in pint pandas... @hgrecco Maybe we could integrate this with your https://github.com/hgrecco/pint-geopandas ?

**Comment by redhog** [comment]
@hgrecco I'd be happy to play around. I guess an MVP would be:

* [ ] PintArray / pint dtypes for non geographic columns in GoeDataFrame
* [ ] GeoDataFrame(df) to retain pint dtypes

?

And I'd add the functionality of my code to that:

* [ ] Pint units for POINT in different CRS
* [ ] Pint delta unit for POINT that can be converted to length
* [ ] Conversion between Series of pint POINT and GeoSeries

**Comment by hgrecco** [comment]
Sounds good to me!

**Comment by redhog** [comment]
@hgrecco My initial tests suggest this "just works", i.e., doing stuff like:

```
>>> a = gpd.GeoDataFrame({"x": [1, 2]}, index=[0, 1], geometry=mygeom).astype({"x": "pint[m]"})
>>> a
   x                    geometry
0  1  POINT (-121.48861 0.00002)
1  2  POINT (-121.48873 0.00001)
>>> a.x.loc[0]
1 meter

```

without any further code! Distance functions don't return pint units though...
